clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -126,7 +128,8 @@ invariant_checker_t::report_if_false(per_shard_t *shard, bool condition,
                   << shard->instr_count_since_last_timestamp_
                   << " instrs since timestamp " << shard->last_timestamp_
                   << "): " << invariant_name << "\n";
-        abort();
+        if (abort_on_invariant_error_)
+            abort();
```

Comment:
> A user would expect the tool to still fail with an error at the end, instead of reporting success. 

Would you expect that to be the case even with -no_abort_on_invariant_error? Not failing in that mode seemed more natural to me since the invariant checker is only counting the errors, which still succeeds even if the trace is bad.

> That user might be us: we run invariant_checker in various tests. 

Tests continue to use the default mode that aborts on any error.

Reply:
Seems like it could go either way: on the one hand our internal version continues, and if it fails I think it might break the infrastructure running it which will assume some unrecoverable error?  OTOH upstream has no such setup and existing users may be relying on the exit code in the abort case.

Reply:
> nd if it fails I think it might break the infrastructure running it which will assume some unrecoverable error

Right.

> OTOH upstream has no such setup and existing users may be relying on the exit code in the abort case.

I don't see the concern though. In the case where the user relies on the exit code, one error is as good as many, so they wouldn't use the -no_abort_on_invariant_error option (note that it is by default set to true) which would simply fail on the first error and avoid wasted work.

Reply:
I think we're mixing up requirements for a user running the invariant checker (who may want to see all errors) and a machine doing the same (as part of a test e.g., use the returned exit code to fail).

Additionally, I don't know if the framework should fail despite the fact that the tool successfully did what it was supposed to do (find invariant errors). I would reserve that behavior for a true error, as in "I couldn't do what I was supposed to because of some fatal error".

To unblock this PR, I can change the exit code to return failure when any invariant error is found. I'll also add a comment detailing this discussion.

Reply:
The ideal end state to me is  flipping the abort_on_invariant_error flag to be false by default, and turn it on only for tests.

Reply:
Hmm so we have 3 use cases?  Interactive user, test suite, and multi-tool framework?  How about a new flag fail_on_invariant_error which fails the tool at the end if it found them, and we turn that off for the framework?  I see the ambiguitiy: how to distinguish a fail b/c it found errors from an internal error type of fail.  I could see just going with your position of not failing and the user has to look for printed messages.  Sorry, don't mean to hold things up, was focusing on interactive user only.

Reply:
Okay so let me keep it as-is for now (though I can change it abort/return fail exit code at the end if you insist). I added a comment noting this.

Reply:
Done.

Reply:
OK.